Adding Performance tests

manolo opened the pull request:
Fixes #713

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-grid/723)
<!-- Reviewable:end -->

limonte commented:
Need to rebase to master, there are a ton of linter errors for some reason.

In future, please make sure there are no linter errors before moving a PR to the Review column.

limonte commented:
I think it makes sense to:

- update the `lighhouse` dependency to the latest (`^1.5.2`) to have the result of actual state of plugin 
- exclude the `gulp perf:run` from `.travis.yml` for now and recheck manually in near future with newer version of `lighhouse`.

Saulis commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[test/performance/config.json, line 2 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecANnraUdsp7LPl3b8:-KecANnraUdsp7LPl3b9:b95js90) ([raw file](https://github.com/vaadin/vaadin-grid/blob/9d3e8188bbf42116d8fb77534337c4113b9dbd73/test/performance/config.json#L2)):*
> ```JSON
> [
>   {"name": "few-inline-items.html", "threshold": 0.5}
> ```

Why we don't run all tests? and should the `threshold` be larger? (Travis seemed to be running somewhere around `0.87`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723)*
<!-- Sent from Reviewable.io -->

manolo commented:
I didn't know that gulpfile in grid was different, thought all were the same for all elements.
PR are moved to the review automatically by waffle.

Added the `'!demo/x-data-provider.html'` exception to the `gulpfile.js`, isn't it a better way to maintain this excepction?

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/performance/config.json, line 2 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecANnraUdsp7LPl3b8:-KecniuJ6XiT5qw8iMRB:b-6vpsk6) ([raw file](https://github.com/vaadin/vaadin-grid/blob/9d3e8188bbf42116d8fb77534337c4113b9dbd73/test/performance/config.json#L2)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Why we don't run all tests? and should the `threshold` be larger? (Travis seemed to be running somewhere around `0.87`
</blockquote></details>

In some git squash I forgot to reenable them.
There are two tests that should be disabled because results are not reliable at all. Added a flag to disable it in the config.json file.
I leave these tests here because they should be useful when lighthouse implementation of time-to-first-meaningful-paint improves.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-:-KdRNADEcJB4G2wzcueD:b-pp66ot)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Lots of helper element imports are missing from the test pages

---

Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[test/performance/few-inline-items.html, line 16 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvCiT0r4_b_7rsHQ8:-KecvCiUgRjIiHy6nQ08:b1esod8) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/few-inline-items.html#L16)):*
> ```HTML
>         font-weight: 400;
>       }
>       .shell {
> ```

Why are these shell styles repeated 5 times in this PR? And what is this "Performance Shell" actually? Skeleton doesn't have it

---

*[test/performance/large-array-items.html, line 139 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvLMNZ527UnVM0cq0:-KecvLMNZ527UnVM0cq1:bpv2rii) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/large-array-items.html#L139)):*
> ```HTML
>     <script>
>      document.querySelector('vaadin-grid').items =
>        [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199];
> ```

Can't this array be generated instead?

---

*[test/performance/scrolling.html, line 138 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-Kecvz23nxXIoNb89pG2:-Kecvz23nxXIoNb89pG3:bdnbaby) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/scrolling.html#L138)):*
> ```HTML
>         var grid = document.querySelector('vaadin-grid');
>         grid.$.scroller.scrollToScaledIndex(60);
>         Polymer.dom.flush();
> ```

Scrolling shouldn't cause any pending dom-operations. These `flush()` calls are not needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-:-Kecx9ornnuwOQtIXv33:b-bzgz8j)*
<!-- Sent from Reviewable.io -->

manolo commented:
this PR is previous to removal of helpers, will add them

---

Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[test/performance/few-inline-items.html, line 16 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvCiT0r4_b_7rsHQ8:-Ked56_pm5hMYdZgwx5Q:b-6vr5wg) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/few-inline-items.html#L16)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Why are these shell styles repeated 5 times in this PR? And what is this "Performance Shell" actually? Skeleton doesn't have it
</blockquote></details>

Adding some layout components to the page improves the performance, it's just a workaround.
Repeated in order not to add an additional import, and styled in a sense manner to fit a real SPI application.

---

*[test/performance/large-array-items.html, line 139 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvLMNZ527UnVM0cq0:-Ked5ZBGyy0HLFa2xJeI:b581j4f) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/large-array-items.html#L139)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Can't this array be generated instead?
</blockquote></details>

I already tried a `for` loop and a `var arr = new Array(200); arr.fill(1);`, both makes lighthouse score worst. So this seems the faster way to add data to the grid to layout the content without penalising score.

---

*[test/performance/scrolling.html, line 138 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-Kecvz23nxXIoNb89pG2:-Ked6PijCcNi4h-VXGBB:b-xeq8x0) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/scrolling.html#L138)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Scrolling shouldn't cause any pending dom-operations. These `flush()` calls are not needed.
</blockquote></details>

Added just for be sure, will remove them so.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-:-Ked6YkZL2Ej2xk4QE48:b-xyl9xo)*
<!-- Sent from Reviewable.io -->

manolo commented:
Added missing imports.

---

Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[test/performance/scrolling.html, line 138 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-Kecvz23nxXIoNb89pG2:-KeeioHOBtl5ZUJydq9u:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/scrolling.html#L138)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Added just for be sure, will remove them so.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-:-KeepI3I-1ZAROu3WXv5:bvs8sxr)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[test/performance/few-inline-items.html, line 16 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvCiT0r4_b_7rsHQ8:-KegnJvQKbnwgtiTy1g7:ba897pe) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/few-inline-items.html#L16)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Adding some layout components to the page improves the performance, it's just a workaround.
Repeated in order not to add an additional import, and styled in a sense manner to fit a real SPI application.
</blockquote></details>

I’m not following. Why does a performance test require an app layout with a dedicated theme and even responsive layout?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[test/performance/few-inline-items.html, line 16 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvCiT0r4_b_7rsHQ8:-KegoqAxGEWSciBKu9M3:b-e76eb7) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/few-inline-items.html#L16)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

I’m not following. Why does a performance test require an app layout with a dedicated theme and even responsive layout?
</blockquote></details>

The theme or the responsive is not required, Just more layout elements in the DOM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[test/performance/few-inline-items.html, line 16 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvCiT0r4_b_7rsHQ8:-Kegp3UWHOLtklbuXa0Z:bz9jpld) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/few-inline-items.html#L16)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

The theme or the responsive is not required, Just more layout elements in the DOM
</blockquote></details>

If wrapping the grid increases the performance in these tests then would a simple unstyled div be enough then?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[test/performance/few-inline-items.html, line 16 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvCiT0r4_b_7rsHQ8:-KeltVLp1quFRSSHFgY6:bmgqwrr) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/few-inline-items.html#L16)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

If wrapping the grid increases the performance in these tests then would a simple unstyled div be enough then?
</blockquote></details>

I don't have any good reason on having those div styled, just to look a real app shell.
Thinking a little bit more, I could move the common styles to a separate .css, probably shouldn't hit the performance so much, or either remove all not layout related properties, what you consider is better for unblocking this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[test/performance/few-inline-items.html, line 16 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723#-KecvCiT0r4_b_7rsHQ8:-Kem5HnFBuv26OxR8VSm:bt4lh9f) ([raw file](https://github.com/vaadin/vaadin-grid/blob/32f41f4ce62b3836091dd5262272ec997c7e23b1/test/performance/few-inline-items.html#L16)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I don't have any good reason on having those div styled, just to look a real app shell.
Thinking a little bit more, I could move the common styles to a separate .css, probably shouldn't hit the performance so much, or either remove all not layout related properties, what you consider is better for unblocking this.

</blockquote></details>

External css is sure better than repeating them in every file. I still don’t fully understand why the performance test layout should be themed though. It’s not our public demo or anything, it’s just used to measure spin-up performance for vaadin-grid. And since other components' performance tests don't have such an app shell, why should vaadin-grid be an exception...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/723)*
<!-- Sent from Reviewable.io -->

manolo commented:
Remove this PR from the queue, but leave the branch.
